### DISCUSSION TITLE
call vba function from twinbasic

### DISCUSSION THREAD

**Discussion by michaelleewebb** [twinbasic/twinbasic#1732] [Q&A] [answered]
i'm new here, i'm following the sample 4. from the samples tab. this works.  when you press the "hello world" button a msgbox appears, i would like to change that. i would like to run a public sub routine in a vba module. is that possible ?? maybe something like "call mysub". my sub routine only writes to the immediate window, no values need to be passed ether way.

**Comment by fafalone** [comment]
I believe you could use the `Application.Run()` function for this... the addin example sets the variable `vbe` to the Application object;

something like `myAddIn.GetVBE().Run "SubName"`  ?

Some source suggests it's a bit more complicated... like `myAddIn.GetVBE().Workbooks("filename.xlsm").Application.Run("SubName")`

Perhaps one of our resident VBA experts can chime in, I've never made VBA addins before, just doing some googling and poking around the sample code.

**Comment by michaelleewebb** [comment]
i was able to try:
myAddIn.GetVBE().run "listfields"
it built, but when i click the button on the addin i get the following error:
run-timer error -2147352570
unknown name

**Reply by Greedquest** [reply]
Is `listfields` a public Sub with no arguments? Is it in a standard module and not a document object (I don't know access, but e.g. not in ThisWorkbook or Sheet1 in Excel)?

**Comment by michaelleewebb** [comment]
also, i'm using msaccess not excel.

**Comment by GCuser99** [comment]
The "application" object in sample 4 is a reference to the VBE, not the Access Application object which has the Run method. Do you need access to the VBE? If not, then sample 5 (MyComAddin) might be more useful...

**Comment by fafalone** [comment]
Would VBE.ActiveProject.Application be a reference to the object model application object?

**Comment by GCuser99** [comment]
> Would VBE.ActiveProject.Application be a reference to the object model application object?
No, I think it returns the VBE object.

**Reply by fafalone** [reply]
So you could write an infinite series of VBE.ActiveProject.Application.ActiveProject.Application.ActiveProject.Application? 🤣 

How *would* you get the application object model?

**Comment by michaelleewebb** [comment]
yes i need access to the vbide. the code is in a standard module. i'm trying different commands but no luck yet.

**Comment by michaelleewebb** [comment]
if i was in a msaccess program and i wanted to run code in another msaccess program i could use the following code. can this be converted to twinbasic ?
Dim appAccess As Access.Application
 'Create instance of Access Application object.
Set appAccess = CreateObject("Access.Application")
 
'Open Test Database in Microsoft Access window.
appAccess.OpenCurrentDatabase "C:\Test\Test.mdb", False
 
'Run Sub procedure.
appAccess.Run "CountTablesInDB"
 
Set appAccess = Nothing

**Comment by GCuser99** [comment]
@fafalone, I don't think you can get the Access Application object from the VBE object (even with an infinite series 😂).

@michaelleewebb, the reason why I asked if you needed the VBE object: If you are not manipulating/managing VBA code with your tB Add-in, but instead wanting to write an Add-in to use the Access Object Model, including running one of your subs stored in a VBA project like in your example above, then look at Sample 5 (MyComAddin). I see nothing in your example code above that would require a VBE add-in like in Sample 4. The Application object in Sample 5 is the Access Application object that you are looking for.

That aside... Although I haven't tried it, I don't see any reason why your code above would not work in tB (you'll need to set a reference to the Microsoft Access 16.0 Object Library or declare appAccess as a late bound object). One thing I would caution though - if you are instead using the GetObject function to get a reference to an existing instance of the Application object, you could run into complications if more than one instance of Access is open. Again, using Sample 5 would be the cleanest if you don't need to manipulate/manage code with your Add-in.

PS: You may have to qualify the CountTablsInDB procedure with something like Test.CountTablsInDB if the procedure lives in Test.mdb (see [here](https://learn.microsoft.com/en-us/office/vba/api/access.application.run) for more details)

**Comment by michaelleewebb** [comment]
i build sample 5 and it worked. (i had to change it to office 12 instead of office 16, i'm running msaccess 2007. added this to the click event:
applicationObject.run "ListFields"
that works now too. but the button is on the ribbon of the msaccess form not the ide. maybe i should give more details of what i'm doing. when i'm coding i would like to have a button that shows be all the fields of a table in the immediate window. what i have now the button is on the msaccess form and if i'm in full screen i would need to min the ide and bring up the msaccess form, press the button then min the msaccess form to get back to the ide. sample 5 is perfect for me except that the button is in the wrong place. so maybe i need to merge sample 4 and 5 together ?

**Comment by fafalone** [comment]
This looks promising: https://stackoverflow.com/questions/39089985/find-out-the-host-application-in-which-add-in-was-activated

```vba
pp As Excel.Application
Set app = _vbe.ActiveVBProject.VBComponents("ThisWorkbook").Properties("Parent").Object`
```

Is there an equivalent for Access?

It looks like RubberDuck uses a genericized version of that:

```cs
protected HostApplicationBase(VBE vbe, string applicationName)
        {
            _applicationName = applicationName;

            try
            {
                var appProperty = vbe.VBProjects
                    .Cast<VBProject>()
                    .Where(project => project.Protection == vbext_ProjectProtection.vbext_pp_none)
                    .SelectMany(project => project.VBComponents.Cast<VBComponent>())
                    .Where(component => component.Type == vbext_ComponentType.vbext_ct_Document
                    && component.Properties.Count > 1)
                    .SelectMany(component => component.Properties.OfType<Property>())
                    .FirstOrDefault(property => property.Name == "Application");
                if (appProperty != null)
                {
                    Application = (TApplication)appProperty.Object;
                }
                else
                {
                    Application = (TApplication)Marshal.GetActiveObject(applicationName + ".Application");
                }
                    
            }
            catch (COMException)
            {
                Application = null; // unit tests don't need it anyway.
            }
        }
```

I'm not up enough on my C# to give you an exact equivalent of how it's walking VBComponents though.

**Comment by GCuser99** [comment, marked as answer]
@michaelleewebb, if you are going down the Sample 5 (MyComAddin) path, then since you have a reference to the Application object, you also have a reference to the VBE object (Application.VBE). That allows you to build the button or menu item in the VBIDE. See one perturbation of code below. The code dispenses with the IRibbon button and instead builds a menu in the VBIDE. Note that you will need to set references to the Microsoft Office XX.0 Object Model and Microsoft Visual Basic for Applications Extensibility to manage the VBIDE menu. I tested it successfully in Office 365, 64-bit.

```vba
[ ClassId ("8153645C-6460-457C-9073-227C6E4EA5C8") ]
Class MyCOMAddin
 
     ' ---------------------------------------------------------------------------------------
    ' For further instructions, please visit https://www.twinbasic.com/preview.html#sample5
    ' ---------------------------------------------------------------------------------------
    
    'Note from GCuser99:
    'this requires a reference to Microsoft Office XX.0 Object Model
    'and the Microsoft Visual Basic for Applications Extensibility

    Implements IDTExtensibility2

    /*
        PLEASE NOTE: make sure you create a DLL of the correct bitness to match your version of VBA
        (e.g. if you're using the 64-bit version of VBA, make sure you change the 'Active Build' in the twinBASIC panel to 'win64')
    */
    
    /*
    [ WithDispatchForwarding ]
    Implements IRibbonExtensibility
    */
    
    /* 
        The [WithDispatchForwarding] attribute seen above is needed so that late-bound calls
        on the IRibbonExtensibility interface get routed to our MyCOMAddin default interface.
        Without it, events like OnHelloWorldClicked (defined below) will not fire.
    */
    
    Private applicationObject As Object ' Excel.Application, late-bound, but you can add a project reference if you wish
    
    Private vbe As VBIDE.VBE
    Private WithEvents menuItem1Events As VBIDE.CommandBarEvents
    Private isConnected As Boolean

    /* 
        Firstly we implement the IDTExtensibility2 interface members
        We don't do anything particularly interesting here, except for taking a reference
        to the Application object provided to us in the OnConnection event
    */
    Sub OnConnection(ByVal Application As Object, _
                            ByVal ConnectMode As ext_ConnectMode, _
                            ByVal AddInInst As Object, _
                            ByRef custom As Variant()) _
        Implements IDTExtensibility2.OnConnection
        
        'MsgBox "MyCOMAddin.OnConnection"
        Set applicationObject = Application
        
        Set vbe = Application.vbe
        isConnected = True
        CreateVBEMenu()
    End Sub
    
    Sub OnDisconnection(ByVal RemoveMode As ext_DisconnectMode, _
                        ByRef custom As Variant()) _
        Implements IDTExtensibility2.OnDisconnection
        ShutdownAddin
        'MsgBox "MyCOMAddin.OnDisconnection"
    End Sub
    
    Sub OnAddInsUpdate(ByRef custom As Variant()) _
        Implements IDTExtensibility2.OnAddInsUpdate
        
        'MsgBox "MyCOMAddin.OnAddInsUpdate"
    End Sub
    
    Sub OnStartupComplete(ByRef custom As Variant()) _
        Implements IDTExtensibility2.OnStartupComplete
        
        'MsgBox "MyCOMAddin.OnStartupComplete"
    End Sub
    
    Sub OnBeginShutdown(ByRef custom As Variant()) _
        Implements IDTExtensibility2.OnBeginShutdown
        ShutdownAddin
        'MsgBox "MyCOMAddin.OnBeginShutdown"
    End Sub
    
    /* 
        Next we implement IRibbonExtensibility.GetCustomUI which lets us provide
        a ribbon XML string to the host
    */
    
    /*
    Private Function GetCustomUI(ByVal RibbonID As String) As String _
        Implements IRibbonExtensibility.GetCustomUI

        Dim strXML As String
        strXML &= "<customUI xmlns=""http://schemas.microsoft.com/office/2006/01/customui"">" & vbCrLf
        strXML &= "  <ribbon startFromScratch=""false"">" & vbCrLf
        strXML &= "    <tabs>" & vbCrLf
        strXML &= "      <tab id=""tabTest"" label=""twinBASIC Test"">" & vbCrLf
        strXML &= "        <group id=""grpTwin1"" label=""Hello Group"">" & vbCrLf
        strXML &= "          <button id=""btnHello"" size=""large"" label=""Hello World"" getImage=""OnGetHelloWorldButtonImage"" onAction=""OnHelloWorldClicked""/>" & vbCrLf
        strXML &= "        </group>" & vbCrLf
        strXML &= "      </tab>" & vbCrLf
        strXML &= "    </tabs>" & vbCrLf
        strXML &= "  </ribbon>" & vbCrLf
        strXML &= "</customUI>" & vbCrLf
        
        Return strXML
    End Function
    */
    
    /* 
        The XML we provided for the ribbon has a button with an onAction property of 'OnHelloWorldClicked'
        When clicked, that will trigger our event...
    */
    
    /*
    Public Sub OnHelloWorldClicked(Control As IRibbonControl)
        MsgBox "Hello world, from twinBASIC!" & vbCrLf & _
                    "(via " & applicationObject.Name & ")"
        
        ' Here we can use applicationObject (Excel.Application object, or Access.Application etc) 
        '  to interact with the host instance.
    End Sub
    */
    
    /* 
        The XML we provided for the ribbon has a button with a getImage property of 'OnGetHelloWorldButtonImage'
        The host will call this routine to obtain an image to display for the ribbon button...
        We have included twinBASIC_icon.bmp (a 24-bit bitmap image) in the Resources\BITMAP\ folder which you can see in the VS code project folder explorer on the left
    */
    
    /*
    Public Function OnGetHelloWorldButtonImage(Control As IRibbonControl) As IPictureDisp
        Return LoadResPicture("twinBASIC_icon.bmp", vbResBitmap)
    End Function
    */
    
    Private Sub menuItem1Events_Click(ByVal CommandBarControl As Object, _
                ByRef Handled As Boolean, ByRef CancelDefault As Boolean) Handles menuItem1Events.Click
        'this is where you could call your procedure with Application.Run....
        MsgBox "Hello world, from twinBASIC!" & vbCrLf & _
                          "(via " & applicationObject.Name & ")"
        Handled = True
        CancelDefault = False
    End Sub
    
    Private Sub CreateVBEMenu()
        Dim menu As CommandBarPopup
        Dim menuItem1 As CommandBarButton
        
        Set menu = vbe.CommandBars("Menu Bar").Controls.Add(Type:=msoControlPopup)
        With menu
            menu.Caption = "twin&BASIC Tools"
            menu.Tag = "twinBASICTools"
            Set menuItem1 = .Controls.Add(Type:=msoControlButton)
            With menuItem1
                .Caption = "&Hello World"
                Set menuItem1Events = vbe.Events.CommandBarEvents(menuItem1)
            End With
        End With
        Set menuItem1 = Nothing
        Set menu = Nothing
    End Sub
    
    Private Sub RemoveVBEMenu()
        /*
        Dim ctl As Object
        Do
            Set ctl = vbe.CommandBars("Menu Bar").FindControl(Tag:="twinBASICTools")
            If ctl IsNot Nothing Then
                ctl.Delete
            End If
        Loop Until ctl Is Nothing
        */
        'below can replace above loop, after debugging phase is complete
        vbe.CommandBars("Menu Bar").Controls("twinBASIC Tools").Delete
    End Sub
    
    Private Sub ShutdownAddin()
        If isConnected = False Then Exit Sub
        RemoveVBEMenu
        Set vbe = Nothing
        Set applicationObject = Nothing
        isConnected = False
    End Sub
     
End Class
```

**Comment by GCuser99** [comment]
@fafalone, not being a COM expert, I'm not certain what the best strategy is for this one. I thought one of the compelling reasons for using a COM Add-in in general is to safely and easily obtain a reference to the Application object. But in this case @michaelleewebb wants the menu/button in the VBIDE and does not seem to need the VBE object for anything else. 

I'm aware of the Application object "hacks" out there. I use a version of the AccessibleObjectFromWindow "hack" to get the Excel/Access Application object for a twinBASIC ActiveX DLL (see [SeleniumVBA GitHub tB project](https://github.com/GCuser99/SeleniumVBA/tree/main/dev/twinBASIC), tBUtils module, callerApplicationObject procedure). So using a hack like that with Sample 4 VBIDE Add-in should work, but would that be preferable to an arguably simpler COM Add-in that builds the menu in the VBIDE???

**Comment by GCuser99** [comment]
@fafalone, I can't find an Access equivalent to the very cool(!!) Excel trick that you found:
```vba
Dim app As Excel.Application
Set app = vbe.ActiveVBProject.VBComponents("ThisWorkbook").Properties("Parent").Object
```
It works because the class module "ThisWorkbook" always exists in an Excel VBA project, and Application is the parent of the ThisWorkbook object. 

I'm not an Access expert, but I don't see an equivalent situation to leverage in Access. It seems the only modules that exist in an Access VBA project are those that the user inserts.

**Comment by michaelleewebb** [comment]
GCuser99 that did the trick. to the menuItem1Events_Click i added this line:
applicationObject.run "listfields"
works perfect.
i will mark this as solved. also i think i can add some more cool stuff to this and i think i saw a snippet section. i'll add there when complete. thanks so much.

**Comment by Greedquest** [comment]
FYI: If you use the method in https://github.com/Greedquest/vbInvoke then you don't need the application object, you can invoke subs in a host agnostic manner using just the vbe object - e.g.

```vba
'instance of the standard module by name
Dim modFoo As Object = GetExtendedModuleAccessor("moduleName", myAddIn.GetVBE.ActiveVBProject)
modFoo.listfields 'call the listfields method
```
